### PR TITLE
[Yaml] Add --exclude and negatable --parse-tags option to lint:yaml command

### PR DESCRIPTION
**Pull request by christingruber**
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | None
| License       | MIT
| Doc PR        | symfony/symfony-docs#14780

Added a new `--exclude` option to exclude one or more specific files from multiple file list:

```lint:yaml dirname --exclude=/dirname/foo.yml --exclude=/dirname/bar.yml``` 
 
Allow negatable for the parse tags option with `--no-parse-tags` and increase the `symfony/console` dependency to version 5.3

### CODE REVIEW

File: src/Symfony/Component/Yaml/Tests/Command/LintCommandTest.php [hunk 1]
```diff
@@ -142,6 +142,35 @@ public function testCustomTagsError()
...
+    {
+        $this->expectException('RuntimeException');
+        $this->createCommandTester()->execute(['filename' => '.yamllint.yml'], ['decorated' => false]);
+    }
+
+    public function testCorrectLintConfig()
+    {
+        $tester = $this->createCommandTester();
+
+        copy(__DIR__.'/../Fixtures/.yamllint.yml', './.yamllint.yml');
```
Comment (christingruber) [inline src/Symfony/Component/Yaml/Tests/Command/LintCommandTest.php:165]: Find a better way, fails with tests...

File: src/Symfony/Component/Yaml/Command/LintCommand.php [hunk 2]
```diff
@@ -98,12 +110,25 @@ protected function execute(InputInterface $input, OutputInterface $output)
         $this->displayCorrectFiles = $output->isVerbose();
         $flags = $input->getOption('parse-tags') ? Yaml::PARSE_CUSTOM_TAGS : 0;
 
+        if ([$this->lintConfig] === $filenames) {
```
Comment (jderusse) [inline src/Symfony/Component/Yaml/Command/LintCommand.php]: hmm.. this provides an argument to load a config file, but prevent using any files.

What's about adding a new options `-c | --config` and checking if a `.yamllint.yaml` file exist in the project DIR

Or checking if the $filenames contains a single file and this file is a valid yamlint file

Comment (christingruber) [reply src/Symfony/Component/Yaml/Command/LintCommand.php]: > What's about adding a new options `-c | --config`...

I think a new option is a good idea. Maybe with an optional value for a custom config file name `--config=[yamllint.yml]`.

What do you think?

File: src/Symfony/Component/Yaml/Command/LintCommand.php [hunk 3]
```diff
@@ -84,6 +95,7 @@ protected function execute(InputInterface $input, OutputInterface $output)
     {
         $io = new SymfonyStyle($input, $output);
         $filenames = (array) $input->getArgument('filename');
+        $excludes = explode(' ', $input->getOption('exclude'));
```
Comment (xabbuh) [inline src/Symfony/Component/Yaml/Command/LintCommand.php]: We can configure the option instead to accept multiple arguments using `InputOption::VALUE_IS_ARRAY`.

File: src/Symfony/Component/Yaml/Command/LintCommand.php (hunk 2 above)
Comment (xabbuh) [reply src/Symfony/Component/Yaml/Command/LintCommand.php]: I would not provide a default config file, but always require to pass the config filename when this otpion is used.

File: src/Symfony/Component/Yaml/Command/LintCommand.php (hunk 3 above)
Comment (christingruber) [reply src/Symfony/Component/Yaml/Command/LintCommand.php]: Changed :+1:

File: src/Symfony/Component/Yaml/Tests/Command/LintCommandTest.php [hunk 4]
```diff
@@ -142,6 +142,44 @@ public function testCustomTagsError()
...
+  excludes: []
+YAML;
+
+        file_put_contents('.yamllint.yml', $yaml);
+
+        $ret = $tester->execute(['filename' => '.yamllint.yml'], ['decorated' => false]);
+
+        $this->assertSame(0, $ret, 'lint:yaml exits with code 0 in case of success');
+        $this->assertStringContainsString('All 1 YAML files contain valid syntax.', trim($tester->getDisplay()));
+        unlink('.yamllint.yml');
```
Comment (OskarStark) [inline src/Symfony/Component/Yaml/Tests/Command/LintCommandTest.php]: We should use a tearDown() method to ensure the file is deleted after all tests, if one failed. Not sure how symfony handles tests like this

Comment (christingruber) [reply src/Symfony/Component/Yaml/Tests/Command/LintCommandTest.php]: Done, but not final. I will move the path to `sys_get_temp_dir()` with the `--config option`.

Comment (christingruber) [reply src/Symfony/Component/Yaml/Tests/Command/LintCommandTest.php]: There is now the `--config|-c` option without default value. I have also changed the tests, which will use the temp dir for the dummy lint config now.

BTW I have also added a shortcut `-e` for the exclude option.

File: src/Symfony/Component/Yaml/.yamllint.yml.dist [hunk 5]
```diff
@@ -0,0 +1,8 @@
+yaml-lint:
```
Comment (xabbuh) [inline src/Symfony/Component/Yaml/.yamllint.yml.dist]: This file is no longer needed, is it?

**Review by xabbuh — COMMENTED** [review]
I think we should also add a test that handles a call of the command with a config file being passed as well as other options.

File: src/Symfony/Component/Yaml/.yamllint.yml.dist (hunk 5 above)
Comment (christingruber) [reply src/Symfony/Component/Yaml/.yamllint.yml.dist]: Yes I think we can drop the file, an example config is documented.

**Comment by christingruber** [comment]
> I think we should also add a test that handles a call of the command with a config file being passed as well as other options.

I've added a simple mixed test, ~~settings from yaml config will have priority in this case~~

**Comment by christingruber** [comment]
I have changed the load order to allow overrides by console args...

**Comment by christingruber** [comment]
A short overview for the last changes:

I have added some tests which handles mixed arguments from config and options from console. I have also changed the load order to allow overrides by console options. With that it is possible to override the values ​​provided from lint config by console options. @xabbuh

File: src/Symfony/Component/Yaml/Command/LintCommand.php [hunk 6]
```diff
@@ -84,7 +94,31 @@ protected function execute(InputInterface $input, OutputInterface $output)
...
+                $this->format = $lintConfig['format'];
+            }
+            if (false === $flags && isset($lintConfig['parse-tags'])) {
+                $flags = $lintConfig['parse-tags'];
+            }
+            if (!\count($filenames) && isset($lintConfig['includes'])) {
+                $filenames = $lintConfig['includes'];
+            }
+            if (!\count($excludes) && isset($lintConfig['excludes'])) {
+                $excludes = $lintConfig['excludes'];
```
Comment (xabbuh) [inline src/Symfony/Component/Yaml/Command/LintCommand.php]: I think I would merge excludes from config file and CLI options.

File: src/Symfony/Component/Yaml/Command/LintCommand.php [hunk 7]
```diff
@@ -84,7 +94,31 @@ protected function execute(InputInterface $input, OutputInterface $output)
...
+            $lintConfig = $this->parseLintConfig($config);
+
+            if (null === $this->format && isset($lintConfig['format'])) {
+                $this->format = $lintConfig['format'];
+            }
+            if (false === $flags && isset($lintConfig['parse-tags'])) {
+                $flags = $lintConfig['parse-tags'];
+            }
+            if (!\count($filenames) && isset($lintConfig['includes'])) {
+                $filenames = $lintConfig['includes'];
```
Comment (xabbuh) [inline src/Symfony/Component/Yaml/Command/LintCommand.php]: I think I would merge includes from config file and CLI options.

File: src/Symfony/Component/Yaml/Command/LintCommand.php [hunk 8]
```diff
@@ -84,7 +94,31 @@ protected function execute(InputInterface $input, OutputInterface $output)
...
+            if (!file_exists($config)) {
+                throw new RuntimeException(sprintf('Lint config "%s" not found.', $config));
+            }
+
+            $lintConfig = $this->parseLintConfig($config);
+
+            if (null === $this->format && isset($lintConfig['format'])) {
+                $this->format = $lintConfig['format'];
+            }
+            if (false === $flags && isset($lintConfig['parse-tags'])) {
```
Comment (xabbuh) [inline src/Symfony/Component/Yaml/Command/LintCommand.php]: We should probably turn the `--parse-tags` option into a negatable option so that one can override the config using either `--parse-tags` or `--no-parse-tags`.

File: src/Symfony/Component/Yaml/Command/LintCommand.php [hunk 9]
```diff
@@ -113,7 +148,9 @@ protected function execute(InputInterface $input, OutputInterface $output)
             }
 
             foreach ($this->getFiles($filename) as $file) {
-                $filesInfo[] = $this->validate(file_get_contents($file), $flags, $file);
+                if (!\in_array($file, $excludes)) {
```
Comment (xabbuh) [inline src/Symfony/Component/Yaml/Command/LintCommand.php]: please pass `true` as the third argument to `in_array()`

File: src/Symfony/Component/Yaml/CHANGELOG.md [hunk 10]
```diff
@@ -6,6 +6,20 @@ CHANGELOG
 
  * Added `github` format support & autodetection to render errors as annotations
    when running the YAML linter command in a Github Action environment.
+ * Added support for optional YAML lint config from file `lint:yaml --config=.yamllint.yml`.
+
+   ```yml
+   yaml-lint:
```
Comment (xabbuh) [inline src/Symfony/Component/Yaml/CHANGELOG.md]: I am not really sure that we need this top-level entry. IMO this would work as well:

```yaml
format: txt
includes:
  - src/Resources/config/foo.yml
  - src/Resources/config/bar.yml
  - src/Resources/config/path/
excludes:
  - src/Resources/config/path/exclude.yml
```

File: src/Symfony/Component/Yaml/Command/LintCommand.php (hunk 8 above)
Comment (christingruber) [reply src/Symfony/Component/Yaml/Command/LintCommand.php]: The negatable option is only avail with symfony/console 5.3, right?

Comment (stof) [reply src/Symfony/Component/Yaml/Command/LintCommand.php]: which is OK as you add this in symfony/yaml 5.3 too. You can bump the min symfony/console version.

File: src/Symfony/Component/Yaml/CHANGELOG.md (hunk 10 above)
Comment (christingruber) [reply src/Symfony/Component/Yaml/CHANGELOG.md]: Yes it works fine without the top level entry. The only technical reason for the tag is to check if the file has a "valid" lint config:

```php
if (!isset($result['yaml-lint'])) {
    throw new RuntimeException(sprintf('Invalid YAML lint config "%s".', $config));
}
```

And with the tag it is possible to mix the config with other configurations in the same file, not sure if someone need this but it is possible. 

Last but not least, the tag could help other developers to detect the reason for the config file...

For me both is OK. What do you think?

File: src/Symfony/Component/Yaml/Command/LintCommand.php (hunk 8 above)
Comment (christingruber) [reply src/Symfony/Component/Yaml/Command/LintCommand.php]: I'm not sure with the negatable: `--parse-tags` give us `true` and `--no-parse-tags` give us `false`, but the default "without the option" is also `false`. So, currently I can not detect if the user give us the option or not and the console option wins always...

I think we have to set a default `null` for this option to handle the override, but this brings a BC... Am I right?

File: src/Symfony/Component/Yaml/Command/LintCommand.php (hunk 6 above)
Comment (christingruber) [reply src/Symfony/Component/Yaml/Command/LintCommand.php]: OK changed.

File: src/Symfony/Component/Yaml/Command/LintCommand.php (hunk 7 above)
Comment (christingruber) [reply src/Symfony/Component/Yaml/Command/LintCommand.php]: OK changed.

File: src/Symfony/Component/Yaml/Command/LintCommand.php (hunk 9 above)
Comment (christingruber) [reply src/Symfony/Component/Yaml/Command/LintCommand.php]: :+1:  changed.

File: src/Symfony/Component/Yaml/CHANGELOG.md (hunk 10 above)
Comment (xabbuh) [reply src/Symfony/Component/Yaml/CHANGELOG.md]: I think that it will be easier to write a config from scratch if we ditch the top-level `yaml-lint` entry. I would then do a naiv validation which will do a check like if `array_diff(array_keys($config), ['format', 'includes', 'excludes', 'parse-tags'])` returns the empty array.

File: src/Symfony/Component/Yaml/Command/LintCommand.php (hunk 8 above)
Comment (xabbuh) [reply src/Symfony/Component/Yaml/Command/LintCommand.php]: Setting the default value to `null` should be fine. If I don't miss anything, that shouldn't break anything.

Comment (christingruber) [reply src/Symfony/Component/Yaml/Command/LintCommand.php]: OK changed.

File: src/Symfony/Component/Yaml/CHANGELOG.md (hunk 10 above)
Comment (christingruber) [reply src/Symfony/Component/Yaml/CHANGELOG.md]: OK so, I have dropped the top level entry and changed the tests.

**Comment by javiereguiluz** [comment]
I'm divided about this proposal. I love the new options that you added to the command but I'm not sure about providing support for a new config file specifically to run this command.

Just asking: if we keep all the new command options, could a makefile task be an alternative to this config file? You would add the values of all the command options to create a very long command ... but since it's a Makefile, you don't have to actually type or remember the long command, only a short alias. Maybe it's not possible, I don't know.

As a poor version of the Makefile, a simple `lint.sh` or `lint.bat` script in the `bin/` dir of the project could work too to avoid passing all the long option values. Cheers!

**Comment by christingruber** [comment]
@javiereguiluz Thanks for your feedback.

No problem I think we can hold the `--exclude` option and drop the config part. If so, I can refactor and rebase the branch in the next days. OK?

**Comment by javiereguiluz** [comment]
@christingruber thanks ... but maybe we should wait for other comments from @symfony/mergers first. The reason is that in my previous comment I only gave my opinion ... and a single person opinion is not important. Better see what's the general consensus about what to do here. Thanks!

**Review by Nyholm — CHANGES_REQUESTED** [review]
Wow, what a great PR. Well done. I like that you added tests and a good description in the changelog. I also like the implementation. 

I do agree with @javiereguiluz, Let's not add a config file. If we would go the config file route, we must also support different formats and it would make sense to allow other commands to have config files too. It will be a large feature and Im not sure about that route. I think the workarounds mentioned before is better.

I would be happy to see this PR with only the `exclude` feature.

**Comment by chalasr** [comment]
Thanks for your work on this.

> I would be happy to see this PR with only the exclude feature.

Same for me :)

**Comment by OskarStark** [comment]
Hi @christingruber 👋 

Are you open to finish this PR for `5.4` 🎯 ?

Cheers

**Comment by christingruber** [comment]
@OskarStark All or only the exclude option?

**Comment by OskarStark** [comment]
Only the exclude please

File: src/Symfony/Component/Yaml/Tests/Command/LintCommandTest.php [hunk 11]
```diff
@@ -166,7 +176,7 @@ protected function createCommandTester(): CommandTester
     {
         $application = new Application();
         $application->add(new LintCommand());
-        $command = $application->find('lint:yaml');
+        $command = $application->find(LintCommand::getDefaultName());
```
Comment (chalasr) [inline src/Symfony/Component/Yaml/Tests/Command/LintCommandTest.php]: unrelated change should be reverted

**Review by chalasr — APPROVED** [review]

File: src/Symfony/Component/Yaml/CHANGELOG.md [hunk 12]
```diff
@@ -1,6 +1,11 @@
 CHANGELOG
 =========
 
+5.4
+---
+
+* Added new `lint:yaml dirname --exclude=/dirname/foo.yml --exclude=/dirname/bar.yml` option to exclude one or more specific files from multiple file list.
```
Comment (Nyholm) [inline src/Symfony/Component/Yaml/CHANGELOG.md]: Add new….

File: src/Symfony/Component/Yaml/Tests/Command/LintCommandTest.php [hunk 13]
```diff
@@ -142,6 +142,16 @@ public function testCustomTagsError()
         $this->assertSame(1, $ret, 'lint:yaml exits with code 1 in case of error');
     }
 
+    public function testLintWithExclude()
+    {
+        $tester = $this->createCommandTester();
+        $filename = $this->createFile('foo: bar');
+
```
Comment (Nyholm) [inline src/Symfony/Component/Yaml/Tests/Command/LintCommandTest.php:150]: Let’s create and other file that we do parse.

File: src/Symfony/Component/Yaml/composer.json [hunk 14]
```diff
@@ -23,10 +23,10 @@
         "symfony/polyfill-php81": "^1.22"
     },
     "require-dev": {
-        "symfony/console": "^4.4|^5.0|^6.0"
+        "symfony/console": "^5.4"
     },
     "conflict": {
-        "symfony/console": "<4.4"
+        "symfony/console": "<5.4"
```
Comment (Nyholm) [inline src/Symfony/Component/Yaml/composer.json]: Can you tell me why we need to update composer.json? 

Is it because the Negatable? Is that related to the exclude feature?

**Review by Nyholm — CHANGES_REQUESTED** [review]
Great. Thank you for this PR. 

I added some questions. Also, can you make sure the PR description is up to date? It will make it easier to understand this feature after the PR is merged.

File: src/Symfony/Component/Yaml/composer.json (hunk 14 above)
Comment (christingruber) [reply src/Symfony/Component/Yaml/composer.json]: It is for the negatable `InputOption::VALUE_NEGATABLE` option.

Comment (chalasr) [reply src/Symfony/Component/Yaml/composer.json]: Using a negatable option was suggested here https://github.com/symfony/symfony/pull/39641#discussion_r556372547

**Review by chalasr — APPROVED** [review]

**Comment by christingruber** [comment]
> I added some questions. Also, can you make sure the PR description is up to date?...

Hi @Nyholm,

thanks for your Feedback, I updated the MR description / changelog and the composer deps. (See my further comments)

**Review by Nyholm — APPROVED** [review]
Thank you. 

I like this feature and the implementation. I also appreciate that you updated the PR description and that you prepared a doc PR.  

Well done.

File: src/Symfony/Component/Yaml/CHANGELOG.md [hunk 15]
```diff
@@ -1,6 +1,14 @@
 CHANGELOG
 =========
 
+5.4
+---
+
+ * Add new `lint:yaml dirname --exclude=/dirname/foo.yml --exclude=/dirname/bar.yml`
+   option to exclude one or more specific files from multiple file list
+ * Allow negatable for the parse tags option with `--no-parse-tags` and increase
+   the `symfony/console` dependency to version 5.3
```
Comment (OskarStark) [inline src/Symfony/Component/Yaml/CHANGELOG.md]: Not sure we need to mention the increase

**Review by OskarStark — APPROVED** [review]
One minor

File: src/Symfony/Component/Yaml/CHANGELOG.md (hunk 15 above)
Comment (fabpot) [reply src/Symfony/Component/Yaml/CHANGELOG.md]: Indeed

Comment (christingruber) [reply src/Symfony/Component/Yaml/CHANGELOG.md:10]: 👍🏻  removed.

Comment (christingruber) [reply src/Symfony/Component/Yaml/CHANGELOG.md]: OK. Removed. (Give me a short hint if it's the other way around)

File: src/Symfony/Component/Yaml/CHANGELOG.md [hunk 16]
```diff
@@ -1,6 +1,14 @@
 CHANGELOG
 =========
 
+5.4
+---
+
+ * Add new `lint:yaml dirname --exclude=/dirname/foo.yml --exclude=/dirname/bar.yml`
```
Comment (OskarStark) [inline src/Symfony/Component/Yaml/CHANGELOG.md]:
```suggestion
 * Add new `lint:yaml dirname --exclude=/dirname/foo.yaml --exclude=/dirname/bar.yaml`
```

File: src/Symfony/Component/Yaml/Command/LintCommand.php [hunk 17]
```diff
@@ -76,6 +77,10 @@ protected function configure()
   <info>php %command.full_name% dirname</info>
   <info>php %command.full_name% dirname --format=json</info>
 
+You can also exclude one or more specific files:
+
+  <info>php %command.full_name% dirname --exclude="dirname/foo.yml" --exclude="dirname/bar.yml"</info>
```
Comment (OskarStark) [inline src/Symfony/Component/Yaml/Command/LintCommand.php]:
```suggestion
  <info>php %command.full_name% dirname --exclude="dirname/foo.yaml" --exclude="dirname/bar.yaml"</info>
```

File: src/Symfony/Component/Yaml/CHANGELOG.md (hunk 16 above)
Comment (christingruber) [reply src/Symfony/Component/Yaml/CHANGELOG.md]: 👍🏻 good 👀

File: src/Symfony/Component/Yaml/composer.json [hunk 18]
```diff
@@ -23,10 +23,10 @@
         "symfony/polyfill-php81": "^1.22"
     },
     "require-dev": {
-        "symfony/console": "^4.4|^5.0|^6.0"
+        "symfony/console": "^5.3"
```
Comment (ogizanagi) [inline src/Symfony/Component/Yaml/composer.json]: `^6.0` should still be allowed, right?

```suggestion
        "symfony/console": "^5.3|^6.0"
```

**Comment by chalasr** [comment]
Thank you @christingruber.
